### PR TITLE
feat(decider,chain_connector,tests): Worker registration tx tracking [NET-575, NET-538]

### PR DESCRIPTION
**Pull request by kmd-fl**
Notable changes:
1. `register_worker` now returns `RegisterWorkerResult` with `tx_hash` on success and an error message on error.
2.  add the deals on which Decider couldn't get the `tx_hash` to `failed_deals`. **We don't retry to send the message until it returns success since Aqua doesn't support it.**
3. Decider now tracks the txs. If the tx failed (`eth_getTransactionReceipt` return `"status": "0x0"`), Decider adds the deal to `failed_deals` without removing it from `joined_deals` (since it's a pain to implement). If the request received an RPC error, we retry the tx later. On success, we just print OK to logs and forget about the tx (maybe we should put it to the worker-spell, idk). 
4. We preserve info about _all_ transactions. `worker_registration_txs` list contains all transaction requests for all deals (`deal_id` and `tx_hash`), `worker_registration_txs_statuses` contains the list of all known statuses (transaction passed, transaction failed). To get the list of unknown tx statuses we just do `all txs - known txs`, it may take a while.

### CODE REVIEW

File: src/aqua/decider/register_worker.aqua [hunk 1]
```diff
@@ -27,9 +43,57 @@ func register_call() -> string:
...
+                    if tx_status.status == "failed":
+                        deal_log(spell_id, tx.deal_id, ["transaction failed", tx.tx_hash])
+                        store_tx_failed_deal(spell_id, tx.deal_id, tx.worker_id, "transaction failed", ?[tx.tx_hash])
+                    else:
+                        if tx_status.status == "pending":
+                            deal_log(spell_id, tx.deal_id, ["transaction still pending", tx.tx_hash])
+                            store_tx(spell_id, tx)
+                        else:
+                            if tx_status.status == "ok":
+                                deal_log(spell_id, tx.deal_id, ["transaction done, can find receipt by tx hash:", tx.tx_hash])
```
Comment (folex) [inline src/aqua/decider/register_worker.aqua:99]: let's add block number to log here

**Comment by kmd-fl** [comment]
sorry for force-push, messed up a little

File: .fluence/aqua/deals.aqua [hunk 2]
```diff
@@ -0,0 +1,17 @@
+aqua Deals declares *
```
Comment (kmd-fl) [inline .fluence/aqua/deals.aqua:1]: is used without corresponding xor

Comment (kmd-fl) [inline .fluence/aqua/deals.aqua:1]: added accidentally

do we need these files? why don't they in gitignore

Comment (folex) [reply .fluence/aqua/deals.aqua:1]: they're not in git ignore to share deployed deal info

you deploy, I can use. and vice versa.

File: src/aqua/decider/deal_storage.aqua [hunk 3]
```diff
@@ -68,17 +67,74 @@ func get_joined_deals(spell_id: SpellId) -> []JoinedDeal:
...
+    if !result.success:
+        errors <<- result.error
+    else:
+        for str <- result.strings:
+            try:
+                deals <<- JsonFailedDeal.parse(str)
+            catch e:
+                errors <- Json.stringify(["can't parse FaileDeal", e, str])
+    <- deals, errors
+
```
Comment (justprosh) [inline src/aqua/decider/deal_storage.aqua:140]:
```suggestion

```

File: src/aqua/decider/register_worker.aqua [hunk 4]
```diff
@@ -1,12 +1,36 @@
...
+import Spell, StringListValue from "@fluencelabs/spell/spell_service.aqua"
+import Array from "@fluencelabs/aqua-lib/math.aqua"
+import Json from "../fluence/peer.aqua"
+import store_tx_failed_deal from "decider/deal_storage.aqua"
 
-func register_call() -> string:
-    <- "OK"
+alias TxHash: string
+
+const TXS_KEY = "worker_registration_txs"
```
Comment (justprosh) [inline src/aqua/decider/register_worker.aqua:14]: Maybe move these consts (like FAILED_DEALS) and TXS_KEY to a separate file?

Comment (justprosh) [reply src/aqua/decider/register_worker.aqua:14]: So it will be easier to track what we have in KV

File: src/aqua/decider/register_worker.aqua [hunk 5]
```diff
@@ -27,9 +51,55 @@ func register_call() -> string:
...
 -- Remember to track registration failures though, and retry them
 
-func register_worker(chain: ChainInfo, spell_id: SpellId, deal_id: DealId, worker_id: WorkerId, pat_id: []u8):
-    err <- ChainConnector.register_worker(pat_id, worker_id, chain, deal_id)
-    if err == nil:
-        deal_log(spell_id, deal_id, "registered worker")
+func register_worker(chain: ChainInfo, spell_id: SpellId, deal_id: DealId, worker_id: WorkerId, pat_id: []u8) -> bool:
+    result <- ChainConnector.register_worker(pat_id, worker_id, chain, deal_id)
+    is_ok = result.success
+    if result.success:
```
Comment (justprosh) [inline src/aqua/decider/register_worker.aqua:57]:
```suggestion
    if is_ok:
```

File: src/aqua/decider/register_worker.aqua [hunk 6]
```diff
@@ -27,9 +51,55 @@ func register_call() -> string:
...
-        deal_log(spell_id, deal_id, ["error registering worker", err!])
+        for status <- result.results:
+            if !status.success:
+                deal_log(spell_id, status.tx.deal_id, ["couldn't find tx status:", "tx_hash:", status.tx.tx_hash, "error:", status.error ])
+            else:
+                if status.status != "pending":
+                    store_tx_status(spell_id, WorkerTxStatus(tx_info = status.tx, status = status.status))
+                    if status.status == "failed":
+                        store_tx_failed_deal(spell_id, status.tx.deal_id, "transaction failed", ?[status.tx.tx_hash], status.block_number)
+                        deal_log(spell_id, status.tx.deal_id, ["transaction failed, tx_hash:", status.tx.tx_hash, "block_number:", status.block_number])
```
Comment (justprosh) [inline src/aqua/decider/register_worker.aqua:105]:
```suggestion
                        deal_log(spell_id, status.tx.deal_id, ["transaction failed, tx_hash:", status.tx.tx_hash, "block_number:", status.block_number])

```

File: src/aqua/fluence/spell.aqua [hunk 7]
```diff
@@ -45,3 +45,10 @@ func get_counter(spell_id: SpellId) -> u32:
     else:
         result <<- 0
     <- result!
+
+func list_push(spell_id: SpellId, key: string, val: ⊤):
+    Spell spell_id
+    str <- Json.stringify(val)
+    result <- Spell.list_push_string(key, str)
+    if !result.success:
+        spell_log(spell_id, ["can't store value in list by key", key, "error:", result.error])
```
Comment (justprosh) [inline src/aqua/fluence/spell.aqua:54]:
```suggestion
        spell_log(spell_id, ["can't store value in list by key", key, "error:", result.error])

```

File: src/aqua/info.aqua [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import Json from "../fluence/peer.aqua"
+
+func get_all_failed_deals(peerId: string) -> []FailedDeal:
+    on peerId:
+        failed_deals, errors <- get_failed_deals("decider")
+
+    if errors != nil:
+        Console.print(["getting registration failed deals return errors", errors])
+
+    <- failed_deals
```
Comment (justprosh) [inline src/aqua/info.aqua:18]:
```suggestion
    <- failed_deals

```

**Review by justprosh — APPROVED** [review]

**Pull request merged**